### PR TITLE
622 remove dependency on go launch a survey

### PR DESCRIPTION
**Pull request by SamGodwin2**
### What is the context of this PR?
Currently in order to support piping of metadata a change needed to be made to survey-launcher in order to provide metadata to Runner. This PR instead removes the dependencey on survey-launcher and instead uses a new endpoint on the api in order to generate a JWT that can be passed to Runner which will include all the metadata values.  

The new endpoint: 
```
query getquestionnaires($id: ID!) {
  launch(id: $id){
    location
  }
}
```

In order to support metadata however a change had to be made to remove certain metadata keys from being included in the JWT as these would have the potential to break the token. These inappropriate keys also had their defaults removed to help prevent them from being used.

 
### How to review 
Tests pass and the new endpoint can be quired and returns a valid JWT that can be passed to Runner.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -9,13 +9,15 @@ services:
       - db
     volumes:
       - .:/app
-      #- ./node_modules/eq-author-graphql-schema:/app/node_modules/eq-author-graphql-schema
+      - ./node_modules/eq-author-graphql-schema:/app/node_modules/eq-author-graphql-schema
```
Comment (tgandrews) [inline docker-compose.yml]: I presume this is a mistake

**Comment by SamGodwin2** [comment]
[Remove dependency on go-launch-a-survey](https://trello.com/c/FJFyZVEK/622-remove-dependency-on-go-launch-a-survey)

File: utils/jwtGeneration/jwtHelper.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+/* eslint-disable camelcase*/
+const KJUR = require("jsrsasign");
+const fs = require("fs");
+const JSONWebKey = require("json-web-key");
+const jose = require("node-jose");
+const JWK = jose.JWK;
```
Comment (tgandrews) [inline utils/jwtGeneration/jwtHelper.js]: Could 5 - 7 be `const { JWK, JWE } = require("node-jose");`?

File: utils/jwtGeneration/jwtHelper.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  fs.readFileSync("./utils/jwtGeneration/keys.yml", "utf8")
+);
+const keysJson = JSON.parse(JSON.stringify(keysYaml, null, 4));
+
+module.exports.generateToken = function(questionnaireId, metadata) {
+  const signingKeyObject = getKeyByUse(keysJson.keys, "signing");
+  const encryptionKeyObject = getKeyByUse(keysJson.keys, "encryption");
+
+  // Header
+  const oHeader = {
```
Comment (tgandrews) [inline utils/jwtGeneration/jwtHelper.js]: What is the `o` prefix? Is it meant to signify an Object?

File: utils/jwtGeneration/jwtHelper.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    alg: "RS256",
```
Comment (tgandrews) [inline utils/jwtGeneration/jwtHelper.js]: Could we lift this and the value below as constants with nicer names. Especially as `RS256` is reused below.

File: utils/refineMetadata.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable camelcase*/
+const { omit, assign } = require("lodash/fp");
+const uuid = require("uuid/v1");
+const KJUR = require("jsrsasign");
```
Comment (tgandrews) [inline utils/refineMetadata.js]: `KJUR`? I don't know this acronym

File: utils/refineMetadata.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  ru_name: "ESSENTIAL ENTERPRISE LTD",
+  trad_as: "ESSENTIAL ENTERPRISE LTD",
+  eq_id: questionnaireId,
+  collection_exercise_sid: uuid(),
+  period_id: "201605",
+  form_type: questionnaireId,
+  survey_url: `${process.env.PUBLISHER_URL}${questionnaireId}`
+});
+
+module.exports.refineMetadata = (metadata, questionnaireId) => {
```
Comment (tgandrews) [inline utils/refineMetadata.js]: Could we call this "sanitise" or something similar - I'm not sure refine is the best verb to use.

File: utils/jwtGeneration/jwtHelper.js (hunk 3 above)
Comment (SamGodwin2) [reply utils/jwtGeneration/jwtHelper.js]: I think so... It's the variable name the guy who wrote the module uses in his tutorial so I kinda just kept it. Maybe its worth changing that and the KJUR names to something a little clearer.

File: repositories/LauncherRepository.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+const { generateToken } = require("../utils/jwtGeneration/jwtHelper");
+const { assign } = require("lodash");
+
+module.exports.launchASurvey = async function(id, ctx) {
```
Comment (samiwel) [inline repositories/LauncherRepository.js]: I would drop the `A` in this name.

File: utils/jwtGeneration/jwtHelper.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+const getKeyByUse = (json, useCase) =>
+  flow(
+    keys,
+    map(kid => assign(json[kid], { kid })),
+    find(keyObject => keyObject.use === useCase)
+  )(json);
+
+const keysYaml = yaml.safeLoad(
+  fs.readFileSync("./utils/jwtGeneration/keys.yml", "utf8")
```
Comment (samiwel) [inline utils/jwtGeneration/jwtHelper.js]: I think putting this file in the project root will make things easier when we eventually grab this file out of S3.

Comment (samiwel) [reply utils/jwtGeneration/jwtHelper.js]: Also, I think we should introduce another environment variable here too. Runner call it something like `KEYS_FILE`. We should do the same even if we just default to this for now.

File: utils/jwtGeneration/jwtHelper.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  // Payload
+  const oPayload = refineMetadata(metadata, questionnaireId);
+
+  // Sign JWT, password=616161
+  const sHeader = JSON.stringify(oHeader);
+  const sPayload = JSON.stringify(oPayload);
+
+  const prvKey = KJUR.KEYUTIL.getKey(signingKeyObject.value, "digitaleq");
+
+  const sJWT = KJUR.jws.JWS.sign("RS256", sHeader, sPayload, prvKey);
```
Comment (samiwel) [inline utils/jwtGeneration/jwtHelper.js]: I think we can do better with the naming of variables here. `signedJWT`, `signedPayload`, `signingKey` etc. Crypto can be confusing enough we should make our lives easier where we can.

File: utils/jwtGeneration/jwtHelper.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      const token =
+        result.protected +
+        "." +
+        result.recipients[0].encrypted_key +
+        "." +
+        result.iv +
+        "." +
+        result.ciphertext +
+        "." +
+        result.tag;
```
Comment (samiwel) [inline utils/jwtGeneration/jwtHelper.js]: I think we can avoid string concatenation here and use JS template string.

File: utils/jwtGeneration/jwtHelper.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  const oHeader = {
+    alg: "RS256",
+    typ: "JWT",
+    kid: signingKeyObject.kid
+  };
+
+  // Payload
+  const oPayload = refineMetadata(metadata, questionnaireId);
+
+  // Sign JWT, password=616161
```
Comment (samiwel) [inline utils/jwtGeneration/jwtHelper.js]: What is this password in the comment? We should remove.

File: utils/jwtGeneration/jwtHelper.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  const sHeader = JSON.stringify(oHeader);
+  const sPayload = JSON.stringify(oPayload);
+
+  const prvKey = KJUR.KEYUTIL.getKey(signingKeyObject.value, "digitaleq");
```
Comment (samiwel) [inline utils/jwtGeneration/jwtHelper.js]: We should probably extract the passcode into an environment variable.

File: utils/jwtGeneration/keys.yml [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+      MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAt8LZnIhuOdL/BC029GOa
+      JkVUAqgp2PcmbFr2Qwhf/514DUUQ9sKJ1rvwvbmmW2zE8JRtdY3ey0RXGtMn5UZH
+      s8NReHzMxvsmHN4VuaGEnFmPwO821Tkvg0LpKsLkotcw793FD/fut44N2lhpTSW2
+      Sc82uG0p9A+Kud8HCIaWaluosghk9rbMGYDzZQk8cA91GtKJRmIOED4PorB/dexD
+      f37qhuWNQgzyNyTti1DTDUIWyzQQJp926vLbkOip6Fc2R13hOFNETe68Rrw/h3hX
+      EFS17uPFZHsxvm9PFXX9KZMS25ohqbNh97I94LL4o4wybl6LaE6lJEHiD6docD0B
+      6wIDAQAB
+      -----END PUBLIC KEY-----
+
+    version: sr
```
Comment (samiwel) [inline utils/jwtGeneration/keys.yml]: As a separate note. It would be good to document how we generated this file somewhere in confluence if it doesn't already exist.

Comment (SamGodwin2) [reply utils/jwtGeneration/keys.yml]: I doubt it exists, I'll take a look and if not write something up.

File: utils/sanitiseMetadata.test.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+      user_id: "UNKNOWN",
+      case_id: 123,
+      ru_ref: "12346789012A",
+      ru_name: "ESSENTIAL ENTERPRISE LTD",
+      trad_as: "ESSENTIAL ENTERPRISE LTD",
+      eq_id: 1,
+      collection_exercise_sid: 123,
+      period_id: "201605",
+      form_type: 1,
+      survey_url: `${process.env.PUBLISHER_URL}1`
```
Comment (samiwel) [inline utils/sanitiseMetadata.test.js]: Do we need to set the env var for the test? Seems like we should be able to put whatever here?

File: utils/sanitiseMetadata.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  case_id: uuid(),
+  ru_ref: "12346789012A",
+  ru_name: "ESSENTIAL ENTERPRISE LTD",
+  trad_as: "ESSENTIAL ENTERPRISE LTD",
+  eq_id: questionnaireId,
+  collection_exercise_sid: uuid(),
+  period_id: "201605",
+  form_type: questionnaireId,
+  survey_url: `${process.env.PUBLISHER_URL}${questionnaireId}`
+});
```
Comment (samiwel) [inline utils/sanitiseMetadata.js]: Can the survey_url be passed in from above?

File: utils/jwtHelper.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  flow(
+    keys,
+    map(kid => assign(json[kid], { kid })),
+    find(keyObject => keyObject.use === useCase)
+  )(json);
+
+const keysYaml = yaml.safeLoad(
+  fs.readFileSync(`${process.env.KEYS_FILE}`, "utf8")
+);
+const keysJson = JSON.parse(JSON.stringify(keysYaml, null, 4));
```
Comment (samiwel) [inline utils/jwtHelper.js]: No benefit to pretty printing here?

File: utils/sanitiseMetadata.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+  "eq_id",
+  "form_type",
+  "account_service_url"
+]);
+
+const defaultMetadata = (questionnaireId, tokenIssueTime, surveyUrl) => ({
+  tx_id: uuid(),
+  jti: uuid(),
+  iat: tokenIssueTime,
+  exp: tokenIssueTime + 10000,
```
Comment (tgandrews) [inline utils/sanitiseMetadata.js]: Can we extract this `10000` as a constant - I presume its the expiry time but in what unit. Is it 10000 seconds or milliseconds?

File: utils/jwtHelper.js [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  // Header
+  const jwtHeader = JSON.stringify({
+    alg: SIGNING_ALGORITHM,
+    typ: "JWT",
+    kid: signingKeyObject.kid
+  });
+
+  const tokenIssuedTime = KJUR.jws.IntDate.get("now");
+
+  const payload = sanitiseMetadata(
```
Comment (tgandrews) [inline utils/jwtHelper.js]: Could we pass this all down to this function instead of it knowing anything about metadata sanitising. It feels like the rest of the function is token specific and this payload generation isn't. So we would call sanitiseMetadata in the launch middleware

**Comment by tgandrews** [comment]
Could we update the PR description to reflect the design change.

File: utils/jwtHelper.js (hunk 18 above)
Comment (SamGodwin2) [reply utils/jwtHelper.js]: I did have it set-up like that originally however it meant we needed to import the KJUR module both here for the crypto stuff and then also in the launch middleware for the expiry time. I moved it here to try to isolate the crypto imports to one module but understand it makes this function a little less specific. Not sure which approach would be better here?

File: middleware/launch.test.js [hunk 19]
```diff
@@ -0,0 +1,49 @@
+const getLaunchUrl = require("./launch");
+const mockRepository = require("../tests/utils/mockRepository");
+
+let repositories;
+const QUESTIONNAIRE_ID = 1;
+
+describe("status middleware", () => {
```
Comment (samiwel) [inline middleware/launch.test.js]: copy and paste error

File: middleware/launch.test.js [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+  });
+
+  it("should call a redirect with a jwt", async () => {
+    const ctx = { repositories };
+    await getLaunchUrl(ctx)(req, res, next);
+
+    expect(res.redirect).toHaveBeenCalled();
+  });
+
+  it("should call next with an error if a metadata is missing a key", async () => {
```
Comment (samiwel) [inline middleware/launch.test.js]: Probably also worth adding tests describing the behaviour of `null` and "    ". Since keys are null by default?

File: utils/sanitiseMetadata.js [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+  user_id: "UNKNOWN",
+  case_id: uuid(),
+  ru_ref: "12346789012A",
+  ru_name: "ESSENTIAL ENTERPRISE LTD",
+  trad_as: "ESSENTIAL ENTERPRISE LTD",
+  eq_id: questionnaireId,
+  collection_exercise_sid: uuid(),
+  period_id: "201605",
+  form_type: questionnaireId,
+  survey_url: `${surveyUrl}${questionnaireId}?r${Date.now()}`
```
Comment (samiwel) [inline utils/sanitiseMetadata.js]: Wonder if `tokenIssueTime` can be used instead of `Date.now()`. Do they produce the same format of output?

File: keys.yml [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+      pDAlNMCA95hN5/8LFIYsUXL/nCbgY0gsd8K5po9ekZCCnpTh1sr61h9jk24mZUz6
+      uyyq94IrWfIGqSfi4DF/42LKdrPm8kU5DNRR4ZOaU3aQpKMt84KyQXL7ElyDLyPD
+      yj5Hm9xF+6mSPYzJJAItYQKBgHzUZXbzf7ZfK2fwVSAlt68BJDvnzP62Z95Hqgbp
+      hjDThXPbvBXYcGkt1fYzIPZPeOxe6nZv/qGOcEGou4X9nOogpMdC09qprTqw/q/N
+      w9vUI3SaW/jPuzeqZH7Mx1Ajhh8uC/fquK7eMe2Dbi0b2XOeB08atrLyhk3ZEMsL
+      2+IFAoGAUbmo0idyszcarBPPsiEFQY2y1yzHMajs8OkjUzOVLzdiMkr36LF4ojgw
+      UCM9sT0g1i+eTfTcuOEr3dAxcXld8Ffs6INSIplvRMWH1m7wgXMRpPCy74OuxlDQ
+      xwPp/1IVvrMqVgnyS9ezAeE0p9u8zUdZdwHz1UAggwbtHR6IbIA=
+      -----END RSA PRIVATE KEY-----
+    version: rrm
```
Comment (samiwel) [inline keys.yml]: Apparently this version is supposed to be numeric so that we can rotate the keys e.g. v1, v2, v3 etc. I don't think it matters a great deal for now. But just for info.

File: utils/jwtHelper.js [hunk 23]
```diff
@@ -0,0 +1,68 @@
+/* eslint-disable camelcase*/
+const KJUR = require("jsrsasign");
+const fs = require("fs");
+const JSONWebKey = require("json-web-key");
+const { JWK, JWE } = require("node-jose");
+const { find, keys, assign, flow, map } = require("lodash/fp");
+const yaml = require("js-yaml");
+const SIGNING_ALGORITHM = "RS256";
+const keysFile = process.env.KEYS_FILE || "./keys.yml";
+const signingPassword = process.env.SIGNING_PASSWORD || "digitaleq";
```
Comment (samiwel) [inline utils/jwtHelper.js]: As discussed, this can be removed 👍
